Python wrapper classes for all user interfaces

timsaucer opened the pull request:
# Which issue does this PR close?

Closes #749.

 # Rationale for this change

As described the the [issue linked above](https://github.com/apache/datafusion-python/issues/749), some python users find the current approach to exposing the `pyo3` generated functions and classes to be "unpythonic" (their words). This PR attempts to make a first class user interface by adding python documentation and type annotations in the place most python users will look. It has the added benefits of a cleaner `help(datafusion.functions)` and similar calls.

# What changes are included in this PR?

- Exposes some functions to python that were not previously available
- Exposes some classes to python that were not previously available
- Adds wrappers around classes
- Adds wrappers around all functions within `datafusion.functions`
- Adds documentation strings for all classes and functions.
- Adds unit tests for some features that were not covered.
- Adds convenience options, such as passing either a `str` or `pathlib.Path` for paths, updating `select` to accept column names as strings, and performing comparisons against literals.

# Are there any user-facing changes?

- Simplifies the module naming. Instead of `datafusion.functions.functions` it is simply `datafusion.functions`. Similar changes occur for other modules.
- When performing expression comparison, you can pass in either an expression or a literal value. Instead of using `df.filter(col("a") > lit(3))` you can simply do `df.filter(col("a") > 3)`. Any value that can be coerced into a literal is valid for the RHS.
- When doing a `select` operation you can pass in either expressions or column names as a string. Instead of doing `df.select(col("a"), col("b") - col("c"))` you can do `df.select("a", col("b") - col("c"))`.
- `filter` now accepts multiple arguments and will treat them as an AND filter. Instead of doing `df.filter(col("a") > 3).filter(col("b") == "orange")` you can use `df.filter(col("a" > 3, col("b") == "orange")`.
- All commands that accept string paths can also accept `pathlib.Path` objects.
- `plan`, `serde`, `producer`, and `consumer` within `datafusion.substrait` are all marked as deprecated. Please use `Plan`, `Serde`, `Producer`, and `Consumer`, respectively.
- For defining an UDF or UDAF, there is `Volatility` enum available for convenience. The `str` values are still accepted.

Example generated documenation:
<img width="766" alt="Screenshot 2024-07-09 at 8 49 27 AM" src="https://github.com/apache/datafusion-python/assets/24943992/22492ec8-c1eb-4c61-91d2-702ba2989735">

andygrove reviewed (commented):
This looks like a big step forward to me. I am not a Python expert though, so it would be good to get reviews from others.

andygrove commented:
@jdye64 @charlesbluca fyi

max-muoto commented on python/datafusion/udf.py:
```diff
@@ -0,0 +1,62 @@
+    def __init__(
+        self,
+        name: str | None,
+        accumulator: Callable,
```
It would be good to provide to be specific with the generics here and above:

```python
accumulator: Callable[..., Any]
```

I think we could could use generics here though:


```python
_R = TypeVar("_R", bound=pyarrow.DataType)

class AggregateUDF:
    def __init__(
        self,
        name: str | None,
        accumulator: Callable[..., _R],
        input_types: list[pyarrow.DataType],
        return_type: _R,
        state_type: list[pyarrow.DataType],
        volatility: str,
    ) -> None:
        self.udf = df_internal.AggregateUDF(
            name, accumulator, input_types, return_type, state_type, volatility
        )

    def __call__(self, *args: Expr) -> Expr:
        args = [arg.expr for arg in args]
        return Expr(self.udf.__call__(*args))
```

timsaucer replied:
Excellent point!

timsaucer replied:
Changed in #265aeb7b33f6438e6412efac888e3e61302f67a6

max-muoto commented on python/datafusion/substrait.py:
```diff
@@ -15,9 +15,156 @@
+
+class serde:
+    @staticmethod
+    def serialize(sql: str, ctx: SessionContext, path: str) -> None:
```
Would this work on `pathlib.Path`?

timsaucer replied:
Updated to accept path. Not yet resolving this comment until unit test is added.

timsaucer replied:
Unit tests added

max-muoto reviewed (commented):
We can add a `py.typed` file to indicate to type-checkers that library is typed now (per [PEP 561](https://peps.python.org/pep-0561/#packaging-type-information))

max-muoto commented on python/datafusion/record_batch.py:
```diff
@@ -0,0 +1,52 @@
+        next_batch = next(self.rbs)
+        return RecordBatch(next_batch) if next_batch is not None else None
+
+    def __iter__(self) -> RecordBatchStream:
```
`typing_extensions.Self` would be more accurate for the return-type here, if this was sub-classed for example.

timsaucer replied:
I appreciate this review! Changed.

max-muoto commented on python/datafusion/record_batch.py:
```diff
@@ -0,0 +1,52 @@
+
+    def __next__(self) -> RecordBatch | None:
+        next_batch = next(self.rbs)
+        return RecordBatch(next_batch) if next_batch is not None else None
```
Generally `__next__` will just raise a StopIteration as opposed to returning None.

timsaucer replied:
Thanks! This was a mistake that I forgot to go back and fix.

max-muoto commented on python/datafusion/expr.py:
```diff
@@ -15,9 +15,256 @@
+class CaseBuilder:
+    def __init__(self, case_builder: expr_internal.CaseBuilder) -> None:
+        """
+        :param case_builder: Internal object. This constructor is not expected to be used by the end user. Instead use :func:`case` to construct.
```
Perhaps the docstring style should be consistent with the other modules?

timsaucer replied:
Good point. Do you have a preferred style? I have no strong opinions on which style as long as it's consistently applied.

emgeee replied:
Don't have an opinion on a specific style but it looks like ruff can be configured to enforce a specific style, which is probably worth enabling https://docs.astral.sh/ruff/settings/#lint_pydocstyle_convention

max-muoto replied:
Likewise, I think just enforcing some consistent style (regardless of what it is, I usually stick to the Google style guide) would be good.

timsaucer replied:
Works for me. Thanks! I'll work on it.

timsaucer replied:
I've added the google style to the ruff stage. It caught a lot of things, but didn't catch these mixed styles. I'll try to look further tomorrow, but did do a massive update on getting the whitespace and flushing out those with mixing docstrings.

max-muoto commented on python/datafusion/context.py:
```diff
@@ -0,0 +1,1167 @@
+
+
+class SessionConfig:
+    def __init__(self, config_options: dict[str, str] = {}) -> None:
```
It's generally best to avoid mutable defaults: https://docs.astral.sh/ruff/rules/mutable-argument-default/

timsaucer replied:
Thanks! Updaed.

slyons commented on python/datafusion/context.py:
```diff
@@ -0,0 +1,1167 @@
+        self.config_internal = self.config_internal.with_greedy_memory_pool(size)
+        return self
+
+    def with_temp_file_path(self, path: str) -> RuntimeConfig:
```
Can we support `pathlib.Path` for this?

timsaucer replied:
Added, but not resolving this comment yet until unit tests are also added

slyons commented on python/datafusion/context.py:
```diff
@@ -0,0 +1,1167 @@
+    def register_csv(
+        self,
+        name: str,
+        path: str,
```
Same comment as above about `pathlib.Path`

timsaucer replied:
Updated and also unit test coverage.

slyons commented on python/datafusion/context.py:
```diff
@@ -0,0 +1,1167 @@
+    def register_parquet(
+        self,
+        name: str,
+        path: str,
```
Same comment about `pathlib.Path`

timsaucer replied:
Updated and added unit test coverage

slyons commented on python/datafusion/context.py:
```diff
@@ -0,0 +1,1167 @@
+    def register_avro(
+        self,
+        name: str,
+        path: str,
```
Same comment about `pathlib.Path`

timsaucer replied:
Updated

slyons commented on python/datafusion/context.py:
```diff
@@ -0,0 +1,1167 @@
+
+    def read_json(
+        self,
+        path: str,
```
Same comment as above re `pathlib.Path`

timsaucer replied:
Updated and added unit test coverage

slyons commented on python/datafusion/substrait.py:
```diff
@@ -15,9 +15,156 @@
+    from datafusion._internal import LogicalPlan
+
+
+class plan:
```
According to PEP8:

> According to PEP8: Class names must follow CamelCase naming (Each new word is capitalized without underscore)

timsaucer replied:
Added, but also keeping a `plan` version marked as deprecated.

Throne3d commented on python/datafusion/functions.py:
```diff
@@ -15,9 +15,1731 @@
+    return Expr(f.right(string.expr, n.expr))
+
+
+def round(arg: Expr) -> Expr:
+    """
+    Round the argument to the nearest integer.
+    """
+    return Expr(f.round(arg.expr))
```
Looks like this is missing its optional second parameter based on the test failures

```
col_name = 'revenue', col_type = Decimal128Type(decimal128(38, 4))

    def df_selection(col_name, col_type):
        if col_type == pa.float64() or isinstance(col_type, pa.Decimal128Type):
>           return F.round(col(col_name), lit(2)).alias(col_name)
E           TypeError: round() takes 1 positional argument but 2 were given
```

Maybe something like this, similar to rpad?
```suggestion
def round(arg: Expr, decimal_places: Expr | None = None) -> Expr:
    """
    Round the argument to the nearest integer. If decimal_places is specified, rounds to the given number of decimal places.
    """
    decimal_places = decimal_places if decimal_places is not None else Expr.literal(0)
    return Expr(f.round(arg.expr, decimal_places.expr))
```
?

Throne3d replied:
In these situations, does it make sense to specify `None` as the default value, or embed `Expr.literal(0)` directly? (Is Expr mutable - could that cause issues similar to the use of `{}` as a default value above? If not, I think this could help better document the default value that's used, in the type hint directly, rather than just implied in the documentation.)

timsaucer replied:
Updated to your recommendation.

Michael-J-Ward commented on python/datafusion/functions.py:
```diff
@@ -15,9 +15,1731 @@
+    return Expr(f.struct(*args))
+
+
+def named_struct(name_pairs: list[(str, Expr)]) -> Expr:
```
Great improvement to the clunk Rust API for `named_struct`

datapythonista commented:
This looks great in terms of functionality, but I personally don't think it's worth to have wrappers for every DataFusion functionality just for the types and the docs. Afaik PyO3 creates Python docstrings from the Rust documentation of the exposed objects. So things can be documented in the existing Rust functions. And as you mentioned, for types, a `.pyi` file could be used. Besides a couple of other things, this seems to be what this PR provides, and I think things will be much easier to maintain if all this duplication of code in the wrappers doesn't exist.

For the renaming part `datafusion.functions.functions -> datafusion.functions`, I think the intended behavior is what I implemented in #751, and shouldn't be needed.

timsaucer commented:
It's a well made point @datapythonista and maintainability is definitely a concern.

I would argue that going with a `.pyi` file approach is **worse** for maintainability than what I am proposing here. My reasoning is that if there is a change made to an underlying function or parameters on the rust side, then there is no validation that the associated `.pyi` files get updated. With the wrapper functions, the python unit tests _may_ catch those changes.

What I'm really trying to address is a concern from users that the `datafusion-python` project doesn't feel "pythonic". I was hoping this approach would give 3 things

1. Documentation flushed out for all exposed classes and functions. As you correctly point out that could be done on the rust side with pyo3.
2. Type hinting. This cannot be done with pyo3 due to a hard limit on the C api for python. The existing work around is the `.pyi` files. This is important to a lot of users, especially those just starting out.
3. User friendly location for investigating the available functions and classes. A user _can_ find the available functions and classes within python by calling `help(datafusion)` or `dir`. They can also look at the online documentation. However, there are still many who prefer to have the python code laid out for their viewing.

This is my reasoning. If the community prefers to go the route of not adding type hinting (or go with pyi approach) I'm willing to move things over to a different location.

datapythonista commented:
Thanks for the detailed information @timsaucer, and for all the work with this.

I'm not an expert on type hints but it's not immediately clear to me what would every approach validate in a reasonable way. By using wrappers we won't have types for objects that don't exist, which can happen in the pyi file. But other than that, feels like validations should ve mostly via  mypy of the unittests, and should be the same for both approaches, no? Or am I missing something?

For your 3rd point, the PR I referenced in the previous message, #751, should address all the `dir()` calls, which should return the correct members. If I'm not missing anything `help()` would be working with point 1, when the objects have docstrings.

So, for me personally, I fully agree with the goals, but feels like a pyi file and documenting objects in Rust will make things simpler to maintain. I may surely be wrong.

Michael-J-Ward commented:
The wrappers also allow us to offer more pythonic interfaces when the `rust` API is "clunky", at least by `python` standards.

An example of this is how @timsaucer cleaned up the  `named_struct` interface

https://github.com/apache/datafusion-python/blob/79bb196357304126fdd157eb71db472e80a7e7e2/python/datafusion/functions.py#L946-L954

timsaucer commented:
@datapythonista I think your PR is great! It definitely cleans up a problem, but it's not the same one as what I was trying to say.

I know there are many users who look at the repositories to try to understand how a project works. All of this information *can* be found by looking in python using `dir` and such, but many people like to look through the repository to see how all the pieces fit together. For these users, having the user facing python code in one place makes it easier to understand the layout and design. Again, all of this same information can be found in other places. I believe that if we want to see greater adoption we should try to teach people in the way that is most comfortable to them.

Again, this isn't a hill I intend to die on. I want this PR to be helpful.

austin362667 commented:
As per `.pyi` discussion, I think it's relevant and worth mentioning this issue https://github.com/PyO3/pyo3/issues/2454 and also keep track the progress of it.

datapythonista commented:
Thanks for the comments, and sorry if my feedback is not helpful. Just one last comment if you don't mind.

Since the final goal seems to be adoption, and making things easier for Python users, the question that comes to my mind is whether this API wants to be a building block for other projects, or it wants to be a reasonable project for final users.

My previous feedback was based on DataFusion being more for developers than for final users. And for example, nice DataFrame APIs for DataFusion being built as separate projects. If the idea is to make this API reasonable for final users, I think the approach here it makes more sense to me (not sure if I'd wrap everything, but some class surely would need it).

For me, the main things that would make DataFusion as usable as other DataFrame libraries are summarized in this example:

```python
import datafusion
from datafusion import col, lit, functions as f
import pyarrow


# something like this would be implemented internally, so users can call `datafusion.read_*`
def _read_parquet(*args, **kwargs):
    ctx = datafusion.SessionContext()
    return ctx.read_parquet(*args, **kwargs)
datafusion.read_parquet = _read_parquet  # creating an alias of `read_*` functions so users don't need to know about `SessionContext` when the defaults are fine


df = (datafusion.read_parquet("buildings.parquet")
                .filter(  # `.filter()` accepting multiple conditions (which will be an AND) instead of having to use `&` with its operator precedence problems
                    col("is_offplan") == False,
                    col("rooms") >= 2,  # `.lit(2)` not being required, and Python literals working with operators
                )
                .aggregate(
                    [col("area_name_en")],
                    [f.mean(col("has_parking").cast(float))],  # `.cast()` accepting Python types, which would be internally converted to the PyArrow equivalent
                )
                .select(
                    col("area_name_en").alias("Area"),
                    col("AVG(has_parking)").alias("Percentage of buildings with parking"),  # removing the default `?table?` in column names, the column name was "AVG(?table?.has_parking)"
                )
     )
```

Implementing those things and similar ones would be worth the wrappers. I personally don't see it worth for users to have an idea of the project structure browsing the source code (when docs can be provided, and do a better job at that).

timsaucer commented:
@datapythonista I love those suggestions. I'll get on it as soon as I can get through some of the other comments.

kylebarron commented:
👋 I'm also building Python-Rust-Arrow libraries and thought it might be useful to chime in.

Ultimately it's a personal preference, so I can't blame you if you all choose the polars route of duplicating the Python API in Python code, but I personally agree with @datapythonista that a pyi file is less maintenance overhead than duplicating the Python API, and I like having the Python docstring right next to the Rust code implementing it.

I thought it might be helpful to show a library that goes the other route but still has types and nice documentation. [arro3](https://github.com/kylebarron/arro3) is a minimal Python library to represent Arrow data in Python, with the following overview:

- Write a Python docstring on the Rust class or method docstring, [example](https://github.com/kylebarron/arro3/blob/227c505714865064389d626c7fe3d79e8bd315c0/pyo3-arrow/src/array.rs#L155-L168).
- [This does use a pyi file for types](https://github.com/kylebarron/arro3/blob/main/arro3-core/python/arro3/core/_rust.pyi). 
- [Install the Python library into the current environment](https://github.com/kylebarron/arro3/blob/227c505714865064389d626c7fe3d79e8bd315c0/.github/workflows/docs-core.yml#L60) when building the docs
- Build docs using mkdocs and mkdocstrings. It's nice that this will use the pyi file for the deployed docs.
- [Deployed docs contain the original docstring](https://kylebarron.dev/arro3/core/latest/api/array/#arro3.core.Array.from_arrow), with nice formatting (I prefer mkdocs and mkdocstrings over sphinx; I find it a lot simpler and with a better end-user UI)

One downside is that `help()` doesn't include the python types from inside Python, since help doesn't look at the pyi files. (Though IDEs like vscode will still use the pyi files)

timsaucer commented:
@kylebarron I hadn't thought before about splitting up the type annotations in one place and the documentation in another. That's very clever.

Also there was this comment on discord about using `pyi` files:

> mypy has a great tool to validate that stubs are correct https://mypy.readthedocs.io/en/stable/stubtest.html

In my mind for this project the rust code isn't necessarily "closer" to the implementation for the vast majority of what we're doing. This repo is to a large extent just exposing the `datafusion` repository features. So adding documentation here for example:

```
expr_fn!(upper, arg1, "Converts the string to all upper case.");
```

isn't particularly "closer" to the implementation of the `upper` function than putting the document string in a python wrapper.

I guess it comes down to this question: Do we want to have features like in @datapythonista 's example above that we can more easily handle in the python side OR do we want this repository to focus on being python exposed rust functions and classes?

timsaucer commented:
Proposal: To get a better feeling for the consensus of the group, can I get a reaction to this comment:

thumbs up: recommend keep going with python wrappers as in this PR
thumbs down: focus on pyi files instead using @kylebarron 's approach above and add mypy stubtest to CI

timsaucer commented:
@austin362667 Thanks! I've been watching that with a lot of interest. The original issue is over 2 years old and still making slow progress. I've dove into some parts of it myself and realize it's not a simple problem.

datapythonista commented:
I've been doing some research, and I think most of the usability improvements I suggested above I think they should be better done in the Rust/PyO3 code, not in Python.

For example, making Python objects literals by default (e.g. `col("age") + 1` being equivalent to `col("age") + lit(1)`) should be quite easy by implementing a trait with a single function in PyO3. The `expr.cast(float)` would be pretty much the same. In PyO3 there is already a decent amount of Python logic. For example the implementation of dunder methods like `__add__` to support operators like `col("age") + lit(1)`.

I agree with your point that it doesn't make a significant difference to have the docs in Python or in PyO3 since both are wrappers of the actual implementation in DataFusion itself.

So, in my opinion, the docs and the typing validation don't seem to make a big difference whether they are implemented in Python or Rust/PyO3.

The extended logic to make the API more Pythonic, I personally thing keeping it in Rust is an advantage, since there is already a decent amount there, doesn't seem much more complicated to implement them in Rust (at least the ones discussed until now). And I think it will make it complicated to know what belongs to with side Python/Rust if we start implementing some in Python. Unless you have a clear criteria for that (I personally don't).

So, the main thing to me seems to be:
- Having Python wrappers helps users checking the code get an intuition of what's available (personally I think if there are good API docs this is not important, but maybe for other users it does make a difference).
- Not having the Python wrappers means a lot less code to maintain, code that needs to be in sync with the PyO3 API.
- I assume certain features will be faster / easier to implement in Python than in Rust/PyO3, particularly things that are complex. I don't thing we have any identified yet, right? I assume implementing the named struct in PyO3 wouldn't be too complex, no?

simicd commented:
First of all, thanks a lot for putting in all the work and drafting this PR @timsaucer!! 

I installed the package you mentioned @kylebarron to see how it behaves in an IDE. As python dev I'm particularly keen on having docstrings & examples being available when I hover over a function. Unfortunately VS Code doesn't show any docstrings with the Rust docs + .pyi approach (not sure if I need to configure anything special, below screenshots for the out-of-the-box behavior for both arro3 and polars after running pip install).

So if that's indicative I'd much rather prefer the Python wrapper approach. Having documentation visible directly in VS Code was one of the big reasons that drove me to polars a while back when I was evaluating different libraries.


**Example**:
E.g. Rust + .pyi files does not seem to provide any documentation in VS Code (this is also the case for the pyarrow library):
![image](https://github.com/apache/datafusion-python/assets/10134699/c1724a6d-3e47-4b5d-9a3f-89012e62d470)

Polars for instance with its wrapper approach does this:
![image](https://github.com/apache/datafusion-python/assets/10134699/44a800d7-f37c-4d33-982b-3d1b92498df5)

<details>
  <summary>Code sample</summary>

  ```python
  # Define pyarrow array
  import pyarrow as pa
  
  pa_array = pa.array([1, 2, 3, 4, 5])
  
  
  # Use pyarrow array in arro3
  from arro3.core import Array as Array3
  
  a3_array = Array3.from_arrow(pa_array)
  
  print(a3_array)
  
  
  # Use pyarrow array in polars
  import polars as pl
  
  df = pl.from_arrow(pa_array)
  print(df)
  ```
</details>

datapythonista commented:
@simicd if it's easy for you, do you mind checking what documentation is shown for `datafusion.DataFrame`? I don't know too much about VS Code, I don't use it myself, but I think it should simply be displaying the `__doc__` attribute of the hovered object. `datafusion.DataFrame` has the `__doc__` attribute with the content `A PyDataFrame is a representation of a logical plan and an API to compose statements.\nUse it to build a plan and `.collect()` to execute the plan and collect the result.\nThe actual execution of a plan runs natively on Rust and Arrow on a multi-threaded environment.`. If you can see that, then it doesn't make a difference if the documentation is in a Python wrapper or in the existing Rust code.

simicd commented:
Of course, here's what VS Code displays when hovering over it

Dataframe:
![image](https://github.com/apache/datafusion-python/assets/10134699/31e20c15-0612-497f-8ba6-f03022b4e80f)

Any of the functions, e.g. `create_dataframe`:
![image](https://github.com/apache/datafusion-python/assets/10134699/5cf5c7b3-e93d-42ef-8850-90cb299f6b43)

datapythonista commented:
Thanks @simicd, this is very helpful. I was assuming they'd use the `__doc__` attribute, but seems like VS Code is internally using pylance which is static code analyzer and it only works when the docstring is written  below the class or object definition. For pandas that means that half of the docstrings won't work either (I guess that's what you meant when you were talking about Polars).

Not sure if it'd be feasible to automatically generate all the Python wrappers when packaging, with the PyO3 docstrings. That would solve both the docs problem and the maintenance problem. But not sure how complex that could be, and if more difficult to maintain than the wrappers themselves.

timsaucer commented:
Updated all docstrings to match google format and added line length check into pre-commit configuration. The only thing I have left on my TODO list is updating the documentation building to get the cross references working right. I am currently testing mkdocs per @kylebarron recommendation.

Once that is complete, I'll move this from Draft to ready and ask for a re-review.

Thank you everyone for the thoughtful and productive discussion!

kylebarron commented:
> I am currently testing mkdocs per @kylebarron recommendation.

Given that this project is currently using sphinx for documentation, it would be a pretty big change to move from sphinx to mkdocs, including new configuration and converting existing re-structured text files to markdown. Maybe it would be better to discuss moving to mkdocs in another issue? I mentioned that in my workflow as I'm a big fan of mkdocs over sphinx, but this PR is already decently large.

Michael-J-Ward commented on benchmarks/db-benchmark/join-datafusion.py:
```diff
@@ -74,7 +74,8 @@ def ans_shape(batches):
 print(ctx)
 
-# TODO we should be applying projections to these table reads to crete relations of different sizes
+# TODO we should be applying projections to these table reads to crete relations
```
```suggestion
# TODO we should be applying projections to these table reads to create relations
```

timsaucer replied:
Changed

Michael-J-Ward commented on examples/substrait.py:
```diff
@@ -18,16 +18,15 @@
 from datafusion import SessionContext
 from datafusion import substrait as ss
 
+# TODO add user changing interface note to PR that datafusion.substrait.substrait is simplified to datafusion.substrait
```
Still `TODO`?

timsaucer replied:
Done. comment removed

Michael-J-Ward commented on python/datafusion/__init__.py:
```diff
@@ -15,206 +15,74 @@
+"""DataFusion python package.
+
+This is a Python library that binds to Apache Arrow in-memory query engine DataFusion.
+See https://datafusion.apache.org/python/index.html for more information.
```
```suggestion
See https://datafusion.apache.org/python for more information.
```

timsaucer replied:
Changed.

timsaucer commented:
Added a temporary work around for CI issues and added an issue to fix once the upstream is resolved. https://github.com/apache/datafusion-python/issues/763

Michael-J-Ward reviewed (approved):
Excellent!

andygrove reviewed (approved):
This is excellent @timsaucer. Thank you for all of the hard work. 

IMO, having Python wrappers is the most practical way to make sure we offer the Pythonic experience that people want.
